tests/ConfigurationSchemaGenerator.Tests/GeneratorTests.cs

```diff
@@ -51,56 +118,1110 @@ public void ShouldRemoveIndentation(string value, string expected)
+            }
+            """;
+
+        var schema = GenerateSchemaFromCode(source, []);
+
+        AssertIsJson(schema,
+            """
+            {
+              "type": "object",
+              "properties": {
+                "TestComponent": {
+                  "type": "object",
+                  "properties": {
+                    "FalseByDefault": {
+                      "type": "boolean",
+                      "default": false
+                    },
+                    "TrueByDefault": {
+                      "type": "boolean",
+                      "default": true
+                    },
+                    "UnknownDefault": {
+                      "type": "boolean"
+                    }
+                  }
+                }
+              }
+            }
+            """);
+    }
+
+    private static string GenerateSchemaFromCode(string sourceText, IEnumerable<MetadataReference> references)
+    {
+        var sourceSyntaxTree = SyntaxFactory.ParseSyntaxTree(SourceText.From(sourceText));
+        var compilationOptions = new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary).WithNullableContextOptions(NullableContextOptions.Enable);
+        var compilation = CSharpCompilation.Create("Test", [s_implicitUsingsSyntaxTree, s_attributesSyntaxTree, sourceSyntaxTree], s_defaultReferences.AddRange(references), compilationOptions);
+        Assert.Empty(compilation.GetDiagnostics().Where(diagnostic => diagnostic.Severity > DiagnosticSeverity.Hidden));
+
+        var configSchemaInfo = ConfigSchemaGenerator.GetConfigurationSchema(compilation.Assembly);
+        Assert.NotNull(configSchemaInfo);
+
+        var parser = new ConfigurationBindingGenerator.Parser(configSchemaInfo, new KnownTypeSymbols(compilation));
+        var spec = parser.GetSchemaGenerationSpec(CancellationToken.None);
+        Assert.NotNull(spec);
+
+        var emitter = new ConfigSchemaEmitter(spec, compilation);
+        return emitter.GenerateSchema();
+    }
+
+    private static void AssertIsJson(string actual, string expected)
+    {
+        var actualJson = JsonNode.Parse(actual)!;
+        var expectedJson = JsonNode.Parse(expected)!;
+
+        var serializerOptions = new JsonSerializerOptions { WriteIndented = true };
+        var expectedText = expectedJson.ToJsonString(serializerOptions);
+        var actualText = actualJson.ToJsonString(serializerOptions);
 
-        var first = stripedNodes.First();
-        Assert.Equal(expected, first.ToString().Trim());
+        Assert.Equal(expectedText, actualText);
     }
 
     [Fact]
     public void IntegrationTest()
     {
         // the 'refs' folder is populated by PreserveCompilationContext in the .csproj
-        var referenceAssemblies = Directory.GetFiles(Path.Combine(Directory.GetCurrentDirectory(), "refs"), "*.dll", SearchOption.AllDirectories)
+        var referenceAssemblies = Directory.GetFiles(Directory.GetCurrentDirectory(), "*.dll", SearchOption.AllDirectories)
```

Comment:
> For example, leaving out <PreserveCompilationContext>true</PreserveCompilationContext> in the test project fails for reasons I don't fully understand.

Check out https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#preservecompilationcontext for an explanation of what PreserveCompilationContext.

We use this in the tests so we can get all the "ref" assemblies that were used at build-time. We can load the "ref" assemblies up when the test runs and pass them to Roslyn, so Roslyn can return valid information.

Reply:
> Do you understand why that happens?

yes, this happens because not all the "ref" assemblies are available when you remove PreserveCompilationContext.

> The integration test doesn't load reference assemblies from the "ref" directory anymore

This isn't true because the code recurses to all directories under the current directory, which includes the `ref` directory.

Reply:
Thanks for the explanation, makes sense.